src/gt4py/next/iterator/transforms/collapse_tuple.py

```diff
@@ -25,8 +31,72 @@ def _get_tuple_size(type_: type_inference.Type) -> int:
     )
     return len(type_.dtype)
 
+def _is_let(node: ir.Node) -> bool:
+    return isinstance(node, ir.FunCall) and isinstance(node.fun, ir.Lambda)
 
-@dataclass(frozen=True)
+def _is_if_call(node: ir.Expr):
+    return isinstance(node, ir.FunCall) and node.fun == im.ref("if_")
+
+def _with_altered_arg(node: ir.FunCall, arg_idx: int, new_arg: ir.Expr):
+    return ir.FunCall(
+        fun=node.fun,
+        args=[arg if i != arg_idx else new_arg for i, arg in enumerate(node.args)]
+    )
+
+def _is_trivial_make_tuple_call(node: ir.Expr):
+    if not (isinstance(node, ir.FunCall) and node.fun == im.ref("make_tuple")):
+        return False
+    if not all(isinstance(arg, (ir.SymRef, ir.Literal)) or _is_trivial_make_tuple_call(arg) for arg in node.args):
+        return False
+    return True
+
+def nlet(bindings: list[tuple[ir.Sym, str], ir.Expr]):
+    return im.let(*[el for tup in bindings for el in tup])
+
+def _short_hash(val: str) -> str:
+    return hashlib.sha1(val.encode('UTF-8')).hexdigest()[0:6]
+
+@dataclasses.dataclass(frozen=True)
+class CannonicalizeBoundSymbolNames(eve.NodeTranslator):
+    """
+    Given an iterator expression cannonicalize all bound symbol names.
+
+    If two such expression are in the same scope and equal so are their values.
+
+    >>> testee1 = im.lambda_("a")(im.plus("a", "b"))
+    >>> cannonicalized_testee1 = CannonicalizeBoundSymbolNames.apply(testee1)
+    >>> str(cannonicalized_testee1)
+    'λ(_csym_1) → _csym_1 + b'
+
+    >>> testee2 = im.lambda_("c")(im.plus("c", "b"))
+    >>> cannonicalized_testee2 = CannonicalizeBoundSymbolNames.apply(testee2)
+    >>> assert cannonicalized_testee1 == cannonicalized_testee2
+    """
+    _uids: UIDGenerator = dataclasses.field(
+        init=False, repr=False, default_factory=lambda: UIDGenerator(prefix="_csym")
+    )
+
+    @classmethod
+    def apply(cls, node: ir.Expr):
+        return cls().visit(node, sym_map=ChainMap({}))
+
+    def visit_Lambda(self, node: ir.Lambda, *, sym_map: ChainMap):
+        sym_map = sym_map.new_child()
+        for param in node.params:
+            sym_map[str(param.id)] = self._uids.sequential_id()
+
+        return im.lambda_(*sym_map.values())(self.visit(node.expr, sym_map=sym_map))
+
+    def visit_SymRef(self, node: ir.SymRef, *, sym_map: dict[str, str]):
+        return im.ref(sym_map[node.id]) if node.id in sym_map else node
+
+def _is_equal_value_heuristics(a: ir.Expr, b: ir.Expr):
+    """
+    Return true if, bot not only if, two expression (with equal scope) have the same value.
+    """
+    return a == b or (CannonicalizeBoundSymbolNames.apply(a) == CannonicalizeBoundSymbolNames.apply(b))
```

Comment:
Please explain the second condition in more detail, I don't immediately find an example.